src/NuGet.Services.AzureSearch/Owners2AzureSearch/OwnerDataClient.cs

```diff
@@ -0,0 +1,171 @@
+                _logger.LogInformation("Uploading owner changes to {BlobName}.", blobName);
+
+                var blobReference = Container.GetBlobReference(blobName);
+                blobReference.Properties.ContentType = "application/json";
+
+                await blobReference.UploadFromStreamAsync(
+                    stream,
+                    AccessCondition.GenerateIfNotExistsCondition());
+            }
+        }
+
+        public async Task ReplaceLatestIndexedAsync(
+            SortedDictionary<string, SortedSet<string>> newData,
+            IAccessCondition accessCondition)
+        {
+            using (var stream = new MemoryStream())
+            {
+                using (var streamWriter = new StreamWriter(
+                    stream,
+                    encoding: new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true),
+                    bufferSize: 1024,
+                    leaveOpen: true))
+                using (var jsonTextWriter = new JsonTextWriter(streamWriter))
+                {
+                    Serializer.Serialize(jsonTextWriter, newData);
+                }
+
+                stream.Position = 0;
+
+                var blobName = GetLatestIndexedBlobName();
+                _logger.LogInformation("Replacing the latest indexed owners from {BlobName}.", blobName);
+
+                var mappedAccessCondition = new AccessCondition
+                {
+                    IfNoneMatchETag = accessCondition.IfNoneMatchETag,
+                    IfMatchETag = accessCondition.IfMatchETag,
+                };
+
+                var blobReference = Container.GetBlobReference(blobName);
+                blobReference.Properties.ContentType = "application/json";
+
+                await blobReference.UploadFromStreamAsync(
+                    stream,
+                    mappedAccessCondition);
+            }
+        }
+
+        private static void ReadStream(Stream stream, Action<string, string> add)
+        {
+            using (var textReader = new StreamReader(stream))
+            using (var jsonReader = new JsonTextReader(textReader))
+            {
+                Guard.Assert(jsonReader.Read(), "The blob should be readable.");
+                Guard.Assert(jsonReader.TokenType == JsonToken.StartObject, "The first token should be the start of an object.");
+                Guard.Assert(jsonReader.Read(), "There should be a second token.");
+                while (jsonReader.TokenType == JsonToken.PropertyName)
+                {
+                    var id = (string)jsonReader.Value;
+
+                    Guard.Assert(jsonReader.Read(), "There should be a token after the property name.");
+                    Guard.Assert(jsonReader.TokenType == JsonToken.StartArray, "The token after the property name should be the start of an object.");
+
+                    var owners = Serializer.Deserialize<List<string>>(jsonReader);
+                    foreach (var owner in owners)
```

Comment:
Thoughts on using `Add(string id, IEnumerable<string> owners)` here to dedup the foreach?

Reply:
Yup, sounds good.